src/vim9execute.c

```diff
@@ -4077,9 +4094,17 @@ exec_instructions(ectx_T *ectx)
 			case EXPR_GEQUAL: res = arg1 >= arg2; break;
 			case EXPR_SMALLER: res = arg1 < arg2; break;
 			case EXPR_SEQUAL: res = arg1 <= arg2; break;
+			case EXPR_LSHIFT: res = (uvarnumber_T)arg1 << arg2;
+					  break;
+			case EXPR_RSHIFT: res = (uvarnumber_T)arg1 >> arg2;
+					  break;
 			default: break;
 		    }
 
+		    if (iptr->isn_arg.op.op_type == EXPR_RSHIFT)
+			// clear the topmost sign bit
+			res &= ~((uvarnumber_T)1 << MAX_LSHIFT_BITS);
```

Comment:
This fixup only works if arg2 is `1`. Something like following is needed.
```
result &= (1 << arg2) - 1 << NBITS - arg2
```
If right shifted by 3 and NBITS is 8 then this should be `&= 1110_0000`.

If you can count on the compiler to respect unsigned and make sure zeros are shifted in, then you don't need a fixup. If you can't count on the compiler, this might be more understandable and doesn't need a fixup.
```
res = arg1 >> 1;
res &= ~(1 << MAX_LSHIFT_BITS);
res = res >> arg2 - 1;
```
None of this is tested, simulated on paper... (_I looked up the precedence against +/-_)

Reply:
Oops, updated to add the missing `~` and `&= 0001_1111`